src/main/java/bio/terra/cli/service/WorkspaceManagerService.java

```diff
@@ -955,4 +854,102 @@ private static boolean isRetryable(Exception ex) {
+   *     thrown by the WSM client or the retries
+   */
+  private void callWithRetries(
+      HttpUtils.RunnableWithCheckedException<ApiException> makeRequest, String errorMsg) {
+    handleClientExceptions(
+        () -> HttpUtils.callWithRetries(makeRequest, WorkspaceManagerService::isRetryable),
+        errorMsg);
+  }
+
+  /**
+   * Execute a function that includes hitting WSM endpoints. Retry if the function throws an {@link
+   * #isRetryable} exception. If an exception is thrown by the WSM client or the retries, make sure
+   * the HTTP status code and error message are logged.
+   *
+   * @param makeRequest function with a return value
+   * @param errorMsg error message for the the {@link SystemException} that wraps any exceptions
+   *     thrown by the WSM client or the retries
+   */
+  private <T> T callWithRetries(
+      HttpUtils.SupplierWithCheckedException<T, ApiException> makeRequest, String errorMsg) {
+    return handleClientExceptions(
+        () -> HttpUtils.callWithRetries(makeRequest, WorkspaceManagerService::isRetryable),
+        errorMsg);
+  }
+
+  /**
+   * Execute a function that includes hitting WSM endpoints. If an exception is thrown by the WSM
+   * client or the retries, make sure the HTTP status code and error message are logged.
+   *
+   * @param makeRequest function with no return value
+   * @param errorMsg error message for the the {@link SystemException} that wraps any exceptions
+   *     thrown by the WSM client or the retries
+   */
+  private void handleClientExceptions(
+      HttpUtils.RunnableWithCheckedException<ApiException> makeRequest, String errorMsg) {
+    handleClientExceptions(
+        () -> {
+          makeRequest.run();
+          return null;
+        },
+        errorMsg);
+  }
+
+  /**
+   * Execute a function that includes hitting WSM endpoints. If an exception is thrown by the WSM
+   * client or the retries, make sure the HTTP status code and error message are logged.
+   *
+   * @param makeRequest function with a return value
+   * @param errorMsg error message for the the {@link SystemException} that wraps any exceptions
+   *     thrown by the WSM client or the retries
+   */
+  private <T> T handleClientExceptions(
+      HttpUtils.SupplierWithCheckedException<T, ApiException> makeRequest, String errorMsg) {
+    try {
+      return makeRequest.makeRequest();
+    } catch (ApiException | InterruptedException ex) {
+      // if this is a WSM client exception, check for a message in the response body
+      if (ex instanceof ApiException) {
+        ApiException apiEx = (ApiException) ex;
+        logger.error(
+            "WSM exception status code: {}, response body: {}, message: {}",
+            apiEx.getCode(),
+            apiEx.getResponseBody(),
+            apiEx.getMessage());
+
+        // try to deserialize the response body into an ErrorReport
+        String apiExMsg;
+        try {
```

Comment:
this code seems redundant with what's in SamService. Maybe consolidate?

Reply:
Frustratingly, the code is very similar and the classes have similar or identical names, but they are not the same classes. (e.g. `ApiException` is not the same class in WSM client library and SAM client library). If the classes at least extended a common base class, that would probably open this up to sharing logic, but unfortunately our client libraries have nothing in common.